### PR TITLE
[ROM] Double ROM size to 16KB

### PR DESCRIPTION
**Pull request by tjaychen**
- Current software development is running out of space, so the ROM is doubled
  to accommodate the increase

- Note, at the moment bram_load.mmi and placement.xdc need to be manually updated.  It seems simple enough to re-create this file as long as a tcl hook can be provided as part of the build flow.  Need to check to see if this is possible.

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
hi all, 
please have a look.

There are a few things about this flow that isn't great... right now both `bram_load.mmi` and `placement.xdc` need to be manually updated when we increase the ROM size.  I've actually seen an example in another project where this can be created on the fly after routing, so as long as can put a hook there (which I think we already can do), seems like we should be able to create a custom tcl command to create this. 

I'll give that a try as a separate PR (or if anyone with more tcl chops than mine wants to do it, that's fine too).

@mcy 
I haven't tested this yet, other than to check that existing verilator stuff didn't break.  Do you mind giving this a whirl to see if it solves your issue.

Also, are my updates to the verilator readme correct? I got the impression that `all` would build for fpga directly, so i updated it to be specific, let me know if that wasn't the right thing to do.

File: doc/ug/getting_started_verilator.md [hunk 1]
```diff
@@ -25,13 +25,14 @@ There are 3 memory types: ROM, RAM and Flash.
...
-For that purpose compile the demo program with "simulation" settings, which adjusts the frequencies to better match the simulation speed. 
+For that purpose compile the demo program with "simulation" settings, which adjusts the frequencies to better match the simulation speed.
 For more information on building software targets refer to the [Software Getting Started Guide]({{< relref "getting_started_sw.md" >}}).
 
 ```console
 $ cd $REPO_TOP
 $ ./meson_init.sh
-$ ninja -C build-out all
+$ ninja -C build-out sw/device/boot_rom/boot_rom_export_sim_verilator
+$ ninja -C build-out sw/device/examples/hello_world/hello_world_export_sim_verilator
```
Comment (imphil) [inline doc/ug/getting_started_verilator.md]: That change is not related to the PR description, please remove it or place it in a separate commit.

Apart from that, I'm not sure I like this change: the old command was easy to remember, the new command is almost impossible to remember. It doesn't matter if it builds slightly too much.

File: doc/ug/getting_started_verilator.md [hunk 2]
```diff
@@ -40,8 +41,8 @@ The programs listed after `--meminit` are loaded into the system's specified mem
 ```console
 $ cd $REPO_TOP
 $ build/lowrisc_systems_top_earlgrey_verilator_0.1/sim-verilator/Vtop_earlgrey_verilator \
-  --meminit=rom,build-bin/sw/device/boot_rom/boot_rom_sim_verilator.elf \
-  --meminit=flash,build-bin/sw/device/examples/hello_world/hello_world_sim_verilator.elf
+  --meminit=rom,build-out/sw/device/boot_rom/boot_rom_sim_verilator.elf \
+  --meminit=flash,build-out/sw/device/examples/hello_world/hello_world_sim_verilator.elf
```
Comment (imphil) [inline doc/ug/getting_started_verilator.md]: Why that change? `build-bin` is the output directory, `build-out` is the temporary build directory. Nothing should rely on a path in `build-bin`.

**Comment by mcy** [comment]
@tjaychen Can confirm that this fixes my issue.

I'm curious though... is there a reason we're going up to 16KB, specifically? Is the *actual* size of the mask rom decided yet?

File: doc/ug/getting_started_verilator.md (hunk 1 above)
Comment (gkelly) [reply doc/ug/getting_started_verilator.md]: It also builds quite quickly, so I think I'm good leaving this as the simpler command for the sake of readability.

**Comment by tjaychen** [comment]
i think this is my misunderstanding. For some reason i thought `all` only
built the fpga versions, i now realize it probably builds both fpga and
verilator.
Let me change this back.

**Comment by tjaychen** [comment]
> * LGTM for the code to increase the size.
> * Please put the changes to the documentation (if necessary at all) in a separate commit; they are not related to the size increase.
> * We specify 8kB of ROM in many places throughout the docs, e.g. https://docs.opentitan.org/hw/top_earlgrey/doc/. Changes are needed there as well.
> 
> In more general terms: It seems that 8k of ROM was a very deliberate decision. @sjgitty is it OK to change the ROM size, or do we need to live with 8k of ROM and squeeze the SW accordingly?

@sjgitty should probably confirm, but i think starting off with a smaller ROM was more or less to make sure we didn't use up all of the FPGA brams too quickly (i think we're >50% now).  My expectation for a final product is that ROM will likely go above 16KB (probably 32KB) as we add defensive code, verification code, and other things such as RNG initialization.

**Comment by tjaychen** [comment]
Addressed comments, please have a look.

**Comment by sriyerg** [comment]
LGTM!

**Comment by sjgitty** [comment]
> In more general terms: It seems that 8k of ROM was a very deliberate decision. @sjgitty is it OK to change the ROM size, or do we need to live with 8k of ROM and squeeze the SW accordingly?

8kB was simply the initial ROM size, not necessarily the final size, and
to be honest, we are far from determining our final ROM code content,
so expect there will be future changes. As @tjaychen mentioned it
was intended to stay small at first, but will grow. We also need to get
some access to an eventual ROM builder to understand how much
flexibility there is in non-power-of-two sizes.

**Comment by imphil** [comment]
FYI @alistair23: I double-checked the tock linker script, and it seems that no adjustments are necessary there (https://github.com/tock/tock/blob/master/boards/opentitan/layout.ld#L3). But it could very well be that I looked in the wrong place.

**Comment by alistair23** [comment]
Tock doesn't run from the ROM so I don't think it should impact us (unless other addresses are changed). I'm not sure why they call it `rom` in the linker script though, it's just flash.

**Comment by openhardwarefan** [comment]
@imphil @tjaychen I think in the bram_load_mmi,  the following line 

  <AddressSpace Name="brom" Begin="0" End="4095">
should be 
  <AddressSpace Name="brom" Begin="0" End="**16383**">
otherwise, splice script will generate a bad bit file.

**Comment by tjaychen** [comment]
sorry what was the line you were trying to point out?

**Comment by openhardwarefan** [comment]
@tjaychen sorry, the 4th line like this works for me:
  <AddressSpace Name="brom" Begin="0" End="16383">

**Comment by tjaychen** [comment]
ah thank you, i will fix that. I've actually be wondering how that parameter should be generated. I'll read up a bit more.

**Comment by tjaychen** [comment]
does it also work for you if it's just an arbitrarily large value?
just curious.

**Comment by tjaychen** [comment]
@openhardwarefan 
please have a look at #1618 , i'm trying to just automatically dump mmi.
Whatever I'm doing isn't very smart, and won't work for ROM beyond a certain size, but I will probably try to address that in the future.  I think should would also fix your problem.

**Comment by openhardwarefan** [comment]
> does it also work for you if it's just an arbitrarily large value?
> just curious.

Sorry, I didn't try it. I will try it and let you know that.

**Comment by openhardwarefan** [comment]
Just tested it, it works, thank you for fixing this.

**Comment by tjaychen** [comment]
@openhardwarefan do you mean you tried #1618?
if yes, thank you!

**Comment by openhardwarefan** [comment]
@tjaychen  Yes.

**Comment by tjaychen** [comment]
thank you!
